src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -174,40 +160,40 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                     {
                         // If there's no razor code in this app, don't do anything.
                         RazorSourceGeneratorEventSource.Log.DiscoverTagHelpersFromReferencesStop();
-                        return ImmutableArray<TagHelperDescriptor>.Empty;
+                        return null;
+                    }
+
+                    // If there aren't any references, we can bail out early.
+                    // Note: compilation.References just enuemrators compilation.ExternalReferences and compilation.DirectiveReferences.
+                    if (compilation.ExternalReferences is [] && compilation.DirectiveReferences is [])
+                    {
+                        return null;
                     }
 
                     var tagHelperFeature = GetStaticTagHelperFeature(compilation);
 
-                    List<TagHelperDescriptor> descriptors = new();
+                    // Typically a project with Razor files will have maany tag helpers in references/
+                    // So, we start with a larger capacity to avoid extra array copies.
+                    var results = new List<TagHelperDescriptor>(capacity: 128);
+
                     foreach (var reference in compilation.References)
                     {
                         if (compilation.GetAssemblyOrModuleSymbol(reference) is IAssemblySymbol assembly)
                         {
-                            descriptors.AddRange(tagHelperFeature.GetDescriptors(assembly));
+                            tagHelperFeature.GetDescriptors(assembly, results);
```

Comment:
>results [](http://example.com/codeflow?start=70&length=7)

Talked in person, but just adding to the PR for posterity. This adds a possibility that an ITagHelperDescriptorProvider can clear out previous ITagHelperDescriptorProvider results. Sounds like this isn't a concern. (If it was, you could just have another list defined in here that you passed in and add it's contents to results after the GetDescriptors call)

Reply:
Good point! I don't think this something we're worried about though. None of the `ITagHelperDescriptorProvider` implementations in the compiler are so nefarious, and the source generator picks a specific implementation to use.